File: src/execution/runtimePermissionsEngine.ts
/**
 * Converted from Python: core/execution/runtime_permissions_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

let _SCOPES: any = ["browser", "native", "filesystem", "connector", "terminal", "vm"];
export function buildRuntimePermissions(scopes: any = null): any {
  var active: any = py.sorted(py.toSet(py.or2(scopes, () => (["browser", "native"]))));
  return {"scopes": Object.fromEntries(py.iter(_SCOPES).map((scope: any) => ([scope, py.contains(active, scope)] as [any, any]))), "active_scopes": active, "bounded": true};
}
export function validateRuntimePermissions(permissions: any, runtime: any, action_type: any): any {
  var scopes: any = py.get(permissions, "scopes", {});
  var runtime_scope: any = (py.contains(_SCOPES, runtime) ? runtime : "browser");
  if ((py.truthy(py.startswith(action_type, "terminal_")) || py.eq(action_type, "terminal_command"))) {
    runtime_scope = "terminal";
  } else if (py.truthy(py.startswith(action_type, "native_"))) {
    runtime_scope = "native";
  } else if (py.truthy(py.startswith(action_type, "vm_"))) {
    runtime_scope = "vm";
  } else if (py.truthy(py.startswith(action_type, "connector_"))) {
    runtime_scope = "connector";
  }
  var allowed: any = py.truthy(py.get(scopes, runtime_scope, false));
  return {"allowed": allowed, "scope": runtime_scope, "deterministic": true, "bounded": true};
}

File: lib/src/execution/runtime_permissions_engine.dart
/// Port of core/execution/runtime_permissions_engine.
const List<String> _scopes = <String>[
  'browser',
  'native',
  'filesystem',
  'connector',
  'terminal',
  'vm',
];

Map<String, dynamic> buildRuntimePermissions({List<String>? scopes}) {
  final List<String> active = (<String>{
    ...(scopes ?? <String>['browser', 'native'])
  }.toList())
    ..sort();
  final Map<String, dynamic> scopeMap = <String, dynamic>{};
  for (final String scope in _scopes) {
    scopeMap[scope] = active.contains(scope);
  }
  return <String, dynamic>{
    'scopes': scopeMap,
    'active_scopes': active,
    'bounded': true,
  };
}

Map<String, dynamic> validateRuntimePermissions(
  Map<String, dynamic> permissions,
  String runtime,
  String actionType,
) {
  final Map<String, dynamic> scopes =
      (permissions['scopes'] as Map<String, dynamic>?) ?? <String, dynamic>{};
  String runtimeScope = _scopes.contains(runtime) ? runtime : 'browser';

  if (actionType.startsWith('terminal_') || actionType == 'terminal_command') {
    runtimeScope = 'terminal';
  } else if (actionType.startsWith('native_')) {
    runtimeScope = 'native';
  } else if (actionType.startsWith('vm_')) {
    runtimeScope = 'vm';
  } else if (actionType.startsWith('connector_')) {
    runtimeScope = 'connector';
  }

  final bool allowed = scopes[runtimeScope] == true;
  return <String, dynamic>{
    'allowed': allowed,
    'scope': runtimeScope,
    'deterministic': true,
    'bounded': true,
  };
}
